### PR TITLE
program-test: API to get individual program info and ELF

### PR DESCRIPTION
**Pull request by ilya-bobyr**
There is a subsequent change where we need to get accounts for a specific program, along with the ELF bytes.

Plus, for upgradable programs, extracting ELF from the account data is not completely trivial.  So if the caller needs it, it is convenient to be able to retrieve it directly.

### CODE REVIEW

File: program-test/src/programs.rs [hunk 1]
```diff
@@ -21,76 +26,198 @@ mod spl_associated_token_account {
...
+}
+
+fn upgradable_program_account(
+    rent: &Rent,
+    ProgramInfo {
+        program_id,
+        loader_id,
+        elf,
+    }: &ProgramInfo,
+    last_modified_slot: u64,
```
Comment (buffalojoec) [inline program-test/src/programs.rs:194]: Looks like you're always setting this to `0`, so why not just set it to `0` within the helper, and don't pass it?
Maybe same with `upgrade_authority_address`?

File: program-test/src/programs.rs [hunk 2]
```diff
@@ -21,76 +26,198 @@ mod spl_associated_token_account {
...
+    /// Upgradable programs, ones that use the [`bpf_loader_upgradeable`] loader, use two accounts.
+    ///
+    /// ELF data is stored in the `program_data_account`, but it is prefixed by an additional
+    /// header.  See [`UpgradeableLoaderState::ProgramData`].
+    Upgradable {
+        main_id: Pubkey,
+        main_account: AccountSharedData,
+        program_data_id: Pubkey,
+        program_data_account: AccountSharedData,
+    },
```
Comment (buffalojoec) [inline program-test/src/programs.rs:87]: nit: I think we should stick to the naming conventions used in the upgradeable loader, ie.
- `program_address` and `program_data_address`
- `program_account` and `program_data_account`
- `Upgradeable`

File: program-test/src/programs.rs [hunk 3]
```diff
@@ -21,76 +26,198 @@ mod spl_associated_token_account {
...
+
+/// Returns one or more accounts for the specified program, and the program ELF.
+///
+/// For non-upgradable programs, `elf` data will be the same as the one stored in the program
+/// account.  But for the upgradable programs the relationship is a bit more complex and it could be
+/// convenient to retrieve the ELF bytes directly.
+pub fn spl_program(
+    rent: &Rent,
+    target_id: Pubkey,
+) -> Result<(Vec<u8>, ProgramAccounts), ProgramLoadError> {
```
Comment (buffalojoec) [inline program-test/src/programs.rs:115]: What do you think about flipping the tuple around and calling this `get_spl_program_with_elf_bytes`?

File: program-test/src/programs.rs [hunk 4]
```diff
@@ -21,76 +26,198 @@ mod spl_associated_token_account {
...
+
+fn program_accounts(rent: &Rent, info: &ProgramInfo) -> ProgramAccounts {
+    let ProgramInfo {
+        program_id,
+        loader_id,
+        elf,
+    } = info;
+
+    match *loader_id {
+        bpf_loader::ID => {
```
Comment (buffalojoec) [inline program-test/src/programs.rs:148]: The previous implementation was keying on only the upgradeable loader ID. I'm wondering if we should do the same thing here. It might make things easier now that BPF Loader is deprecated and Loader v4 is being rolled out soon. Also it would avoid that panic at the end. What do you think?

**Review by buffalojoec — COMMENTED** [review]
Looking good! I left a few suggestions on the API.

File: program-test/src/programs.rs (hunk 4 above)
Comment (ilya-bobyr) [reply program-test/src/programs.rs:148]: This suggestion seems a bit contradictory to me :P

A v4 loader program account is not the same as a v2 program account.
So, if we compare `loader_id` only with `bpf_loader_upgradeable::ID` we are not going to support v4 programs.
We would just do the wrong thing if anyone sets the `loader_id` to anything but `bpf_loader_upgradeable::ID` or `bpf_loader::ID`.
I do not see how this makes things simpler %)

And just in general, I think, the code is easier to read when cases are explicit.
Rather than when different branches are merged together, and nested `if` statements are used for the distinct parts.
In this particular case, as we now return different enum branches, the code would get pretty hard to follow if we try to combine different cases.

Actually, it might be better to use  `V2`, and `V3` (and, later, `V4`) as the branch names in the `ProgramAccounts` enum, rather than `Upgradeable`/`NonUpgradeable`.
I have renamed the branches.

As for the `panic` at the end - I think it is a good check in any case.
As the `loader_id` value could be anything, it would be wrong to continue execution if the program loader is something this function was not written to support.
So, I think, it is better to add the `panic` check.

File: program-test/src/programs.rs (hunk 1 above)
Comment (ilya-bobyr) [reply program-test/src/programs.rs:194]: I think it is bad if a function has a generic name, but then is actually specialized to a certain particular use case.
Making these arguments explicit moves the specialization into the caller, where it is slightly more obvious.
Not perfect, but the caller is already specialized, so, at least, this way all the specialization happens in a single function.
And it is the API of the crate, an entry point - I think, a better place for specialization to be.
Compared to a lower level helper function.

File: program-test/src/programs.rs (hunk 2 above)
Comment (buffalojoec) [reply program-test/src/programs.rs]: I agree we have pretty bad duality on `id` vs. `address`. I was more concerned with introducing the keyword `main`, which I don't see used anywhere else to describe program accounts, especially in the loaders.

File: program-test/src/programs.rs (hunk 1 above)
Comment (buffalojoec) [reply program-test/src/programs.rs:198]: These are both private helpers, though. As per your argument, it would make more sense to provide the specialization on the module's entrypoint(s): `spl_program` and `spl_programs`. Considering their use case(s) that seems like overkill.

Unless of course you plan to export these helpers in the later commit you've mentioned, wherein you'll need to use those inputs?

File: program-test/src/programs.rs (hunk 4 above)
Comment (buffalojoec) [reply program-test/src/programs.rs:152]: Alright that's fair. It seems you're making the argument that the original code would need updating anyway, since it was written before V2 was deprecated and V4 was written. Makes sense to me.

This is really a test-only crate, so it's not as crucial, but why not consider adding this error to the `ProgramLoadError` and making this function return a `Result`?

File: program-test/src/programs.rs (hunk 3 above)
Comment (buffalojoec) [reply program-test/src/programs.rs]: Personally I'm on the side of believing it always makes more sense for the function signature to describe what's going to be returned, especially if it's returning a tuple, where only one element is strictly typed (via struct).

Without mentioning `elf` anywhere, this function appears to return some bytes and a `ProgramAccounts` type. One might infer these bytes are ELF bytes, but it's not explicitly known unless they view the function's source or documentation.

With that being said, I guess the documentation provides enough information for this to be used as-is, if you insist on the simple name!

> What is your reasoning for putting them in the opposite order?

Mostly for the naming convention of `spl_program_with_*`, nothing more.

Comment (CriesofCarrots) [reply program-test/src/programs.rs]: Sorry for swooping in here, but please generally avoid returning opaque tuples (especially the one that has 3 components on L200 😬 )

File: program-test/src/programs.rs (hunk 2 above)
Comment (ilya-bobyr) [reply program-test/src/programs.rs]: Ah, I just realized that I did not rename the `main` part.

Well, I think it is actually another problem that we do not have a separate name for the main account of the program.
Consider what happens if I rename `main_account` to `program_account`.
Data stored in this account is now called `program_account_data`.
But we also have a `program_data_account`.
Which is the account holding the program data.
I find it quite confusing.

If you insist, I will do it.  But I think it reduces readability.

File: program-test/src/programs.rs (hunk 4 above)
Comment (ilya-bobyr) [reply program-test/src/programs.rs:152]: `Result` should be used when the caller can do something meaningful with the error.
For example, if it is something that the caller did incorrectly.
Or if the caller would want to handle the error.

`spl_program` returns a `Result`:

```rust
#[derive(Error, Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProgramLoadError {
    #[error("Unknown program address: {0}")]
    UnknownAddress(Pubkey),
}

pub fn spl_program(
    rent: &Rent,
    target_address: Pubkey,
) -> Result<(ProgramElf, ProgramAccounts), ProgramLoadError> {
```

It is used to indicate to the caller that they are trying to load an unknown program.

But an invalid loader is something internal to this module.
The caller can not do anything about it and is not expected to know that this can even happen.
This is a convention between the definition of

```rust
static SPL_PROGRAMS: &[ProgramInfo] = &[ ... ];
```

and `program_accounts` being able to construct accounts for all the loaders specified in the `SPL_PROGRAMS` array.

So, I do not think that returning an error to the caller in case an unexpected loader was specified is correct here.
There is nothing the caller could do about it.
It would only force all caller to write extra error handing code they do not care about.

File: program-test/src/programs.rs (hunk 3 above)
Comment (buffalojoec) [reply program-test/src/programs.rs]: Nice! With the refactoring and the type alias this is much easier to follow.

File: program-test/src/programs.rs (hunk 2 above)
Comment (buffalojoec) [reply program-test/src/programs.rs]: Yeah, I agree `program_account.data` and `program_data_account.data` aren't the greatest combination, but I think it makes more sense to align with the Loader.

https://github.com/solana-labs/solana/blob/74758d9fbfa811f7c868ad2ecca05ce5db1add25/programs/bpf_loader/src/lib.rs#L710

https://github.com/solana-labs/solana/blob/74758d9fbfa811f7c868ad2ecca05ce5db1add25/programs/bpf_loader/src/lib.rs#L769

I'd much rather use `program_data` over `programdata`, but the latter is technically the current convention, so if that's better in your opinion then it makes sense!

File: program-test/src/programs.rs [hunk 5]
```diff
@@ -21,76 +26,189 @@ mod spl_associated_token_account {
...
-            ));
-            accounts.into_iter()
+        .flat_map(|info| {
+            let accounts = program_accounts(rent, info);
+            Vec::<_>::from(accounts).into_iter()
         })
         .collect()
 }
+
+fn program_accounts(rent: &Rent, info: &ProgramInfo) -> ProgramAccounts {
```
Comment (CriesofCarrots) [inline program-test/src/programs.rs:146]: Nitty, but why not make these internal methods on `ProgramAccounts`? Something like:
```
fn from(rent: &Rent, info: &ProgramInfo) -> Self { .. }
fn loader_v2(rent: &Rent, address: Pubkey, elf: &[u8]) -> Self { .. }
fn loader_v3(rent: &Rent, main_address: Pubkey, elf: &[u8]) -> { .. }
```

File: program-test/src/programs.rs [hunk 6]
```diff
@@ -21,76 +26,189 @@ mod spl_associated_token_account {
...
+
+/// Returns the program ELF and the specified program account(s).
+///
+/// For loader v2, non-upgradable programs, `elf` data will be the same as the one stored in the
+/// program account.  But for the loader v3 and loader 4, upgradable programs, the relationship is a
+/// bit more complex and it could be convenient to retrieve the ELF bytes directly.
+pub fn spl_program(
+    rent: &Rent,
+    target_address: Pubkey,
+) -> Result<(ProgramElf, ProgramAccounts), ProgramLoadError> {
```
Comment (CriesofCarrots) [inline program-test/src/programs.rs:120]: I understand that `fn spl_programs()` predates you, but since this one is a new api, I'd still prefer it did not return a tuple.

File: program-test/src/programs.rs (hunk 2 above)
Comment (CriesofCarrots) [reply program-test/src/programs.rs]: Fwiw, I got hung up on `main` because it has no meaning to me in this context; I tend to agree with Joe here.

File: program-test/src/programs.rs (hunk 5 above)
Comment (ilya-bobyr) [reply program-test/src/programs.rs:146]: We are talking about a module that literally has 5 functions in it.
And most are there to construct the `ProgramAccounts`.
I am not sure renaming `program_accounts_loader_v2` to `ProgramAccounts::loader_v2` changes anything in terms of readability.
It will just add a level of indentation to the function implementation.

File: program-test/src/programs.rs (hunk 6 above)
Comment (ilya-bobyr) [reply program-test/src/programs.rs:120]: Why?
It would help if you would provide more reasoning, as I am not sure what is your motivation here.

It is a two element tuple with types that clearly state what the elements are.
At the call site the values will be named, if the code is properly written.
I see no downsides.

If I add a struct, it will mostly add noise, I think.
Compare:

```rust
let (elf, accounts) = spl_program(rent, spl_token_2022::ID);
```

```rent
let ProgramTestProgram {
    elf,
    accounts
} = spl_program(rent, spl_token_2022::ID);
```

The latter just seems to be more verbose.

And if the caller wants to use two programs, they would need to use explicit renaming, making it even more verbose.

Should there be 3 or 4 fields, with some of the fields being optional at different call sites, then I would certainly introduce structs to name the fields.  This is what is done for things stored in `ProgramAccounts`.

**Comment by ilya-bobyr** [comment]
@CriesofCarrots @buffalojoec
I am working on a subsequent change that might be more aligned with what you both were talking about here.
I want to add more programs to the library, not just those in SPL.
And I want to get ELF without the associated accounts.
But it still seems reasonable to share some of the functionality.
Here is a draft, based on this PR: https://github.com/solana-labs/solana/pull/35314
Maybe considering the direction, you would be more happy with the changes in this PR :)

**Comment by willhickey** [comment]
This repository is no longer in use. Please re-open this pull request in the agave repo: https://github.com/anza-xyz/agave
